Modules/_abc.c

```diff
@@ -0,0 +1,1049 @@
+    return 0;
+}
+
+PyDoc_STRVAR(_reset_caches_doc,
+"Internal ABC helper to reset both caches of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_caches(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_reset_caches", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (_reset_negative_cache(self) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+_get_negative_cache_version(PyObject *self)
+{
+    _abc_data *impl;
+    PyObject *res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+    res = impl->_abc_negative_cache_version;
+    Py_DECREF(impl);
+    return res;
+}
+
+static int
+_set_negative_cache_version(PyObject *self, PyObject *version)
+{
+    _abc_data *impl;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    Py_DECREF(impl->_abc_negative_cache_version);
+    impl->_abc_negative_cache_version = abc_invalidation_counter;
+    Py_INCREF(impl->_abc_negative_cache_version);
+    Py_DECREF(impl);
+    return 0;
+}
+
+PyDoc_STRVAR(_get_dump_doc,
+"Internal ABC helper for cache and registry debugging.\n\
+\n\
+Return shallow copies of registry, of both caches, and\n\
+negative cache version. Don't call this function directly,\n\
+instead use ABC._dump_registry() for a nice repr.");
+
+static PyObject *
+_get_dump(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry, *cache, *negative_cache, *cache_version;
+    PyObject *res = PyTuple_New(4);
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_get_dump", 1, 1, &self)) {
+        return NULL;
+    }
+    registry = _get_registry(self);
+    if (!registry) {
+        return NULL;
+    }
+    registry = PyObject_CallMethod(registry, "copy", NULL);
+    if (!registry) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+    cache = PyObject_CallMethod(impl->_abc_cache, "copy", NULL);
+    if (!cache) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    negative_cache = PyObject_CallMethod(impl->_abc_negative_cache, "copy", NULL);
+    if (!negative_cache) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    cache_version = _get_negative_cache_version(self);
+    if (!cache_version) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_INCREF(registry);
+    Py_INCREF(cache);
+    Py_INCREF(negative_cache);
+    Py_INCREF(cache_version);
+    if ((PyTuple_SetItem(res, 0, registry) < 0) ||
+        (PyTuple_SetItem(res, 1, cache) < 0) ||
+        (PyTuple_SetItem(res, 2, negative_cache) < 0) ||
+        (PyTuple_SetItem(res, 3, cache_version) < 0)) {
```

Comment:
`PyTuple_SetItem` does an INCREF itself.  Do you want to `registry`, `cache`, `negative_cache` and `cache_version` twice here?  If so please add a comment explaining why.